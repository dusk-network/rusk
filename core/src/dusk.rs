// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

//! Dusk denomination.

const DUSK_UNIT: f64 = 1_000_000_000.0;

/// The minimum increment of Dusk.
pub const LUX: Dusk = dusk(1.0 / DUSK_UNIT);

/// The Dusk denomination. Use the [`dusk`] function to convert from floating
/// point format, and the [`from_dusk`] function to convert back to Dusk.
///
/// Values of Dusk should *never* be assigned directly. Instead they should use
/// a call to the [`dusk`] function. If increments of the smallest denomination
/// are desired, the [`LUX`] constant can be used.
pub type Dusk = u64;

/// Converts from floating point format to Dusk.
#[must_use]
#[allow(clippy::cast_possible_truncation)]
#[allow(clippy::cast_sign_loss)]
pub const fn dusk(value: f64) -> Dusk {
    (value * DUSK_UNIT) as Dusk
}

/// Converts from Dusk to floating point format.
#[must_use]
#[allow(clippy::cast_precision_loss)]
pub const fn from_dusk(dusk: Dusk) -> f64 {
    dusk as f64 / DUSK_UNIT
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn from_to_dusk() {
        let value = 5f64;
        let dusk_value = dusk(value);

        assert_eq!(value, from_dusk(dusk_value));
    }

    #[test]
    fn lux_is_one() {
        assert_eq!(LUX, 1);
    }
}
