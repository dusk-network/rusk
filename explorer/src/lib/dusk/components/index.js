export { default as Anchor } from "./anchor/Anchor.svelte";
export { default as AnchorButton } from "./anchor-button/AnchorButton.svelte";
export { default as Banner } from "./banner/Banner.svelte";
export { default as Badge } from "./badge/Badge.svelte";
export { default as Button } from "./button/Button.svelte";
export { default as Card } from "./card/Card.svelte";
export { default as CopyButton } from "./copy-button/CopyButton.svelte";
export { default as Icon } from "./icon/Icon.svelte";
export { default as NavList } from "./nav-list/NavList.svelte";
export { default as ProgressBar } from "./progress-bar/ProgressBar.svelte";
export { default as Rerender } from "./rerender/Rerender.svelte";
export { default as RelativeTime } from "./relative-time/RelativeTime.svelte";
export { default as Select } from "./select/Select.svelte";
export { default as Switch } from "./switch/Switch.svelte";
export { default as Textbox } from "./textbox/Textbox.svelte";
export { default as Tooltip } from "./tooltip/Tooltip.svelte";
export { default as Toast } from "./toast/Toast.svelte";
