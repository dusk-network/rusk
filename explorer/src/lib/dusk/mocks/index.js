export { default as IntersectionObserver } from "./IntersectionObserver";
export { default as MediaQueryList } from "./MediaQueryList";
export { default as MediaQueryListEvent } from "./MediaQueryListEvent";
