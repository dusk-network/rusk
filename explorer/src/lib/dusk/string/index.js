export { default as calculateAdaptiveCharCount } from "./calculateAdaptiveCharCount";
export { default as decodeHexString } from "./decodeHexString";
export { default as getRelativeTimeString } from "./getRelativeTimeString";
export { default as hexToBytes } from "./hexToBytes";
export { default as isValidHex } from "./isValidHex";
export { default as makeClassName } from "./makeClassName";
export { default as middleEllipsis } from "./middleEllipsis";
