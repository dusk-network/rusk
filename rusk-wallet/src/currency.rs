// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

//! Module to help with currency conversions.

use core::cmp::Ordering;
use std::fmt;
use std::hash::{Hash, Hasher};
use std::num::ParseFloatError;
use std::ops::{Add, Deref, Div, Mul, Sub};
use std::str::FromStr;

use dusk_core::{dusk, from_dusk};

use crate::Error;

/// The underlying unit of Dusk
pub type Lux = u64;

/// Denomination for DUSK
#[derive(Copy, Clone, Debug, Eq)]
pub struct Dusk(Lux);

impl Dusk {
    /// The smallest value that can be represented by Dusk currency
    pub const MIN: Dusk = Dusk(0);
    /// The largest value that can be represented by Dusk currency
    pub const MAX: Dusk = Dusk(dusk(f64::MAX / dusk(1.0) as f64));

    /// Returns a new Dusk based on the [Lux] given
    pub const fn new(lux: Lux) -> Dusk {
        Self(lux)
    }
}

/// Core ops
/// Implementations of Addition, Subtraction, Multiplication,
/// Division, and Comparison operators for Dusk

/// Addition
impl Add for Dusk {
    type Output = Self;
    fn add(self, other: Self) -> Self {
        Self(self.0 + other.0)
    }
}

impl Add<Lux> for Dusk {
    type Output = Self;
    fn add(self, other: Lux) -> Self {
        Self(self.0 + other)
    }
}

/// Subtraction
impl Sub for Dusk {
    type Output = Self;
    fn sub(self, other: Self) -> Self {
        Self(self.0 - other.0)
    }
}

impl Sub<Lux> for Dusk {
    type Output = Self;
    fn sub(self, other: Lux) -> Self {
        Self(self.0 - other)
    }
}

/// Multiplication
impl Mul for Dusk {
    type Output = Self;
    fn mul(self, other: Self) -> Self {
        let a = from_dusk(self.0);
        let b = from_dusk(other.0);
        Self(dusk(a * b))
    }
}

impl Mul<Lux> for Dusk {
    type Output = Self;
    fn mul(self, other: Lux) -> Self {
        let a = from_dusk(self.0);
        let b = from_dusk(other);
        Self(dusk(a * b))
    }
}

/// Division
impl Div for Dusk {
    type Output = Self;
    fn div(self, other: Self) -> Self {
        Self(dusk(self.0 as f64 / other.0 as f64))
    }
}

impl Div<Lux> for Dusk {
    type Output = Self;
    fn div(self, other: Lux) -> Self {
        Self(dusk(self.0 as f64 / other as f64))
    }
}

/// Equality
impl Hash for Dusk {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}
impl PartialEq for Dusk {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}
impl PartialEq<Lux> for Dusk {
    fn eq(&self, other: &Lux) -> bool {
        self.0 == *other
    }
}
impl PartialEq<f64> for Dusk {
    fn eq(&self, other: &f64) -> bool {
        self.0 == dusk(*other)
    }
}

/// Comparison
impl Ord for Dusk {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0.cmp(other)
    }
}

impl PartialOrd for Dusk {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl PartialOrd<Lux> for Dusk {
    fn partial_cmp(&self, other: &Lux) -> Option<Ordering> {
        self.0.partial_cmp(other)
    }
}
impl PartialOrd<f64> for Dusk {
    fn partial_cmp(&self, other: &f64) -> Option<Ordering> {
        self.0.partial_cmp(&dusk(*other))
    }
}

/// Conversion ops
/// Convenient conversion of primitives to and from Dusk

/// Floats are used directly as Dusk value
impl TryFrom<f64> for Dusk {
    type Error = Error;
    fn try_from(val: f64) -> Result<Self, Error> {
        if val < 0.0 {
            return Err(Error::Conversion(
                "Dusk type does not support negative values".to_string(),
            ));
        }
        Ok(Self(dusk(val)))
    }
}

impl From<Dusk> for f64 {
    fn from(val: Dusk) -> f64 {
        from_dusk(*val)
    }
}

impl From<&Dusk> for f64 {
    fn from(val: &Dusk) -> f64 {
        (*val).into()
    }
}

/// Lux represent Dusk in their underlying unit type
impl From<Lux> for Dusk {
    fn from(lux: Lux) -> Self {
        Self(lux)
    }
}

/// Strings are parsed as Dusk values (floats)
impl FromStr for Dusk {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parse_result = f64::from_str(s).map_err(|e: ParseFloatError| {
            Error::Conversion(format!(
                "Failed to parse Dusk from string: {}",
                e
            ))
        })?;

        Dusk::try_from(parse_result)
    }
}

/// Dusk derefs into its underlying Lux amount
impl Deref for Dusk {
    type Target = Lux;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

/// Display
/// Let the user print stuff
impl fmt::Display for Dusk {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let v: f64 = self.into();
        f64::fmt(&v, f)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basics() {
        let one = Dusk::try_from(1.0).unwrap();
        let dec = Dusk::try_from(2.25).unwrap();
        assert_eq!(one, 1.0);
        assert_eq!(dec, 2.25);
        assert_eq!(Dusk::MIN, 0);
        assert_eq!(Dusk::MIN, Dusk::try_from(0.0).unwrap());
    }

    #[test]
    fn compare_dusk() {
        let one = Dusk::try_from(1.0).unwrap();
        let two = Dusk::try_from(2.0).unwrap();
        let dec_a = Dusk::try_from(0.00025).unwrap();
        let dec_b = Dusk::try_from(0.00190).unwrap();
        assert!(one == one);
        assert!(one != two);
        assert!(one < two);
        assert!(one <= two);
        assert!(one >= one);
        assert!(dec_a < dec_b);
        assert!(one > dec_b);
    }

    #[test]
    fn ops_dusk_dusk() {
        let one = Dusk::try_from(1.0).unwrap();
        let two = Dusk::try_from(2.0).unwrap();
        let three = Dusk::try_from(3.0).unwrap();
        assert_eq!(one + two, three);
        assert_eq!(three - two, one);
        assert_eq!(one * one, one);
        assert_eq!(two * one, two);
        assert_eq!(two / one, two);
        let point_five = Dusk::try_from(0.5).unwrap();
        assert_eq!(one / two, point_five);
        assert_eq!(point_five * point_five, Dusk::try_from(0.25).unwrap());
    }

    #[test]
    fn ops_dusk_lux() {
        let one = Dusk::try_from(1.0).unwrap();
        let one_dusk = 1000000000;
        assert_eq!(one + one_dusk, 2.0);
        assert_eq!(one - one_dusk, 0.0);
        assert_eq!(one * one_dusk, 1.0);
        assert_eq!(one / one_dusk, 1.0);
    }

    #[test]
    fn conversions() {
        let my_float = 35.049;
        let dusk: Dusk = my_float.try_into().unwrap();
        assert_eq!(dusk, my_float);
        let one_dusk = 1_000_000_000u64;
        let dusk: Dusk = one_dusk.try_into().unwrap();
        assert_eq!(dusk, 1.0);
        assert_eq!(*dusk, one_dusk);
        let dusk = Dusk::from_str("69.420").unwrap();
        assert_eq!(dusk, 69.420);
        let float: f64 = dusk.try_into().unwrap();
        assert_eq!(float, 69.420);
        let borrowed = &Dusk(one_dusk);
        let float: f64 = borrowed.try_into().unwrap();
        assert_eq!(float, 1.0);
        let zero = 0;
        assert_eq!(Dusk::try_from(zero).unwrap(), 0);
        let zero = 0.0;
        assert_eq!(Dusk::try_from(zero).unwrap(), 0.0);
    }

    #[test]
    #[should_panic]
    fn overflow() {
        let ten = Dusk::try_from(10.0).unwrap();
        let _ = Dusk::MAX + ten;
    }

    #[test]
    fn negative_dusk() {
        assert!(Dusk::try_from(-1.0).is_err());
    }

    #[test]
    #[should_panic]
    fn negative_result() {
        let one = Dusk::try_from(1.0).unwrap();
        let two = Dusk::try_from(2.0).unwrap();
        let _ = one - two;
    }
}
