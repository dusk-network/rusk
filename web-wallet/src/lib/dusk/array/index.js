export { default as arraysEqual } from "./arraysEqual";
export { default as arrayMax } from "./arrayMax";
export { default as arrayMaxByKey } from "./arrayMaxByKey";
export { default as shuffleArray } from "./shuffleArray";
export { default as findFirstNMatches } from "./findFirstNMatches";
