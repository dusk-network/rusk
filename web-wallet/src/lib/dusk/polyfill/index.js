/** @overview Import the module to load all polyfills at once */

import "./asyncIterator";
import "./promiseWithResolvers";

// eslint-disable-next-line import/no-unresolved
import "web-streams-polyfill/polyfill";
