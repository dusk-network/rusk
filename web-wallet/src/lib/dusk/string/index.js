export { default as calculateAdaptiveCharCount } from "./calculateAdaptiveCharCount";
export { default as cleanNumberString } from "./cleanNumberString";
export { default as hexStringToBytes } from "./hexStringToBytes";
export { default as isValidEvmAddress } from "./isValidEvmAddress";
export { default as makeClassName } from "./makeClassName";
export { default as middleEllipsis } from "./middleEllipsis";
export { default as randomUUID } from "./randomUUID";
export { default as truncateDecimals } from "./truncateDecimals";
