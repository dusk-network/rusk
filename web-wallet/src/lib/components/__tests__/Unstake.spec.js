import {
  afterAll,
  afterEach,
  beforeAll,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { createCurrencyFormatter } from "$lib/dusk/currency";

import { Unstake } from "..";
import { mdiDatabaseArrowDownOutline, mdiGiftOpenOutline } from "@mdi/js";

vi.mock("$lib/dusk/string", async (importOriginal) => {
  /** @type {typeof import("$lib/dusk/string")} */
  const original = await importOriginal();

  return {
    ...original,
    randomUUID: () => "some-generated-id",
  };
});

describe("Unstake", () => {
  const formatter = createCurrencyFormatter("en", "DUSK", 9);
  const lastTxId = "some-id";

  const baseProps = {
    execute: vi.fn().mockResolvedValue(lastTxId),
    formatter,
    gasLimits: {
      gasLimitLower: 10000000n,
      gasLimitUpper: 2900000000n,
      gasPriceLower: 1n,
    },
    gasSettings: {
      gasLimit: 20000000n,
      gasPrice: 1n,
    },
  };

  afterEach(() => {
    cleanup();
    baseProps.execute.mockClear();
  });

  describe("Unstake operation", () => {
    const expectedExplorerLink = `/explorer/transactions/transaction?id=${lastTxId}`;

    const unstakeProps = {
      availableBalance: 10_000_000_000_000n,
      maxWithdrawAmount: 278_000_000_000n,
      minStakeRequirement: 1_000_000_000n,
      operationCtaIconPath: mdiDatabaseArrowDownOutline,
      operationCtaLabel: "Unstake",
      operationOverviewLabel: "Unstake Amount",
    };

    beforeAll(() => {
      vi.useFakeTimers();
    });

    afterAll(() => {
      vi.useRealTimers();
    });

    it("should render the unstake view", () => {
      const baseOptions = {
        props: {
          ...baseProps,
          ...unstakeProps,
        },
        target: document.body,
      };

      const { container, getByRole } = render(Unstake, baseOptions);
      const nextButton = getByRole("button", { name: "Next" });

      expect(nextButton).toBeEnabled();
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should perform an ustake, give a success message and supply a link to see the transaction in the explorer", async () => {
      /** @type {import("svelte").ComponentProps<Unstake>} */
      const props = { ...baseProps, ...unstakeProps };

      const { getByRole, getByText } = render(Unstake, props);

      await vi.advanceTimersToNextTimerAsync();

      await fireEvent.click(getByRole("button", { name: "Next" }));

      await fireEvent.click(getByRole("button", { name: "Unstake" }));

      expect(baseProps.execute).toHaveBeenCalledTimes(1);
      expect(baseProps.execute).toHaveBeenCalledWith(
        unstakeProps.maxWithdrawAmount,
        baseProps.gasSettings.gasPrice,
        baseProps.gasSettings.gasLimit
      );

      const explorerLink = getByRole("link", { name: /explorer/i });

      expect(getByText("Transaction created")).toBeInTheDocument();
      expect(explorerLink).toHaveAttribute("target", "_blank");
      expect(explorerLink).toHaveAttribute("href", expectedExplorerLink);
    });

    it("should not allow to unstake, if wrong gas settings are provided", async () => {
      /** @type {import("svelte").ComponentProps<Unstake>} */
      const props = {
        ...baseProps,
        ...unstakeProps,
        gasSettings: { gasLimit: 29000000090n, gasPrice: 1n },
      };

      const { getByRole } = render(Unstake, props);

      await vi.advanceTimersToNextTimerAsync();

      const nextButton = getByRole("button", { name: "Next" });

      expect(nextButton).toBeDisabled();
    });
  });

  describe("Claim Rewards operation", () => {
    const expectedExplorerLink = `/explorer/transactions/transaction?id=${lastTxId}`;

    const claimRewardsProps = {
      availableBalance: 10_000_000_000_000n,
      maxWithdrawAmount: 345_000_000_000n,
      minStakeRequirement: undefined,
      operationCtaIconPath: mdiGiftOpenOutline,
      operationCtaLabel: "Claim Rewards",
      operationOverviewLabel: "Rewards Amount",
    };

    beforeAll(() => {
      vi.useFakeTimers();
    });

    afterAll(() => {
      vi.useRealTimers();
    });

    it("should render the claim rewards view", () => {
      const baseOptions = {
        props: {
          ...baseProps,
          ...claimRewardsProps,
        },
        target: document.body,
      };

      const { container, getByRole } = render(Unstake, baseOptions);
      const nextButton = getByRole("button", { name: "Next" });

      expect(nextButton).toBeEnabled();
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should perform a claim rewards, give a success message and supply a link to see the transaction in the explorer", async () => {
      /** @type {import("svelte").ComponentProps<Unstake>} */
      const props = { ...baseProps, ...claimRewardsProps };

      const { getByRole, getByText } = render(Unstake, props);

      await vi.advanceTimersToNextTimerAsync();

      await fireEvent.click(getByRole("button", { name: "Next" }));

      await fireEvent.click(getByRole("button", { name: "Claim Rewards" }));

      expect(baseProps.execute).toHaveBeenCalledTimes(1);
      expect(baseProps.execute).toHaveBeenCalledWith(
        claimRewardsProps.maxWithdrawAmount,
        baseProps.gasSettings.gasPrice,
        baseProps.gasSettings.gasLimit
      );

      const explorerLink = getByRole("link", { name: /explorer/i });

      expect(getByText("Transaction created")).toBeInTheDocument();
      expect(explorerLink).toHaveAttribute("target", "_blank");
      expect(explorerLink).toHaveAttribute("href", expectedExplorerLink);
    });

    it("should not allow to claim rewards, if wrong gas settings are provided", async () => {
      /** @type {import("svelte").ComponentProps<Unstake>} */
      const props = {
        ...baseProps,
        ...claimRewardsProps,
        gasSettings: { gasLimit: 29000000090n, gasPrice: 1n },
      };

      const { getByRole } = render(Unstake, props);

      await vi.advanceTimersToNextTimerAsync();

      const nextButton = getByRole("button", { name: "Next" });

      expect(nextButton).toBeDisabled();
    });
  });
});
