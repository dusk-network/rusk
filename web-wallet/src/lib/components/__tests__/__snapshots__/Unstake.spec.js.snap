// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Unstake > Claim Rewards operation > should render the claim rewards view 1`] = `
<div
  class="operation"
>
  <div
    class="dusk-wizard"
  >
    <div
      slot="stepper"
    >
      <div
        class="dusk-stepper dusk-stepper--variant--primary"
        style="--columns: 3; --progress-width: 0%;"
      >
        <span
          aria-current="step"
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step dusk-stepper__step--processed"
        >
          1
        </span>
         
        <span
          class="dusk-stepper__step-label"
          id="step-some-generated-id"
        >
          Amount
           
        </span>
        <span
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step"
        >
          2
        </span>
         
        <span
          class="dusk-stepper__step-label dusk-stepper__step-label--invisible"
          id="step-some-generated-id"
        >
          Overview
           
        </span>
        <span
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step"
        >
          3
        </span>
         
        <span
          class="dusk-stepper__step-label dusk-stepper__step-label--invisible"
          id="step-some-generated-id"
        >
          Done
           
        </span>
        
      </div>
      
    </div>
     
     
     
    <div
      class="operation__amount-wrapper svelte-1p653d4"
    >
      <p>
        Amount:
      </p>
       
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          USE MAX
        </span>
      </button>
    </div>
     
    <div
      class="operation__input-wrapper svelte-1p653d4"
    >
      <input
        class="dusk-textbox dusk-textbox-number operation__input-field"
        max="345"
        required=""
        step="0.000000001"
        type="number"
      />
      
       
      <svg
        class="dusk-icon dusk-icon--size--default"
        data-tooltip-id="main-tooltip"
        data-tooltip-text="DUSK"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
        />
      </svg>
      
    </div>
     
    <div
      class="gas-settings svelte-1daay07"
    >
      <dl
        class="gas-settings__edit svelte-1daay07"
      >
        <div
          class="gas-fee svelte-1c6b481"
        >
          <span>
            Fee:
          </span>
           
          <div
            class="gas-fee__amount svelte-1c6b481"
          >
            <div
              class="gas-fee__amount-value svelte-1c6b481"
            >
              <span
                class="svelte-1c6b481"
              >
                max
              </span>
               
              <span
                class="svelte-1c6b481"
              >
                0.020000000
              </span>
            </div>
             
            <svg
              class="dusk-icon dusk-icon--size--default dusk-amount__icon"
              data-tooltip-id="main-tooltip"
              data-tooltip-text="DUSK"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
              />
            </svg>
            
          </div>
        </div>
        <dd>
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
            type="button"
          >
            <span
              class="dusk-button__text"
            >
              EDIT
            </span>
          </button>
        </dd>
      </dl>
       
    </div>
     
    
     
    <div
      class="dusk-wizard__step-navigation"
    >
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
        type="button"
      >
        <svg
          class="dusk-icon dusk-icon--size--default dusk-button__icon"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
          />
        </svg>
        
         
        <span
          class="dusk-button__text"
        >
          Back
        </span>
        
      </button>
      
       
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          Next
        </span>
         
        <svg
          class="dusk-icon dusk-icon--size--default dusk-button__icon"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
          />
        </svg>
        
      </button>
    </div>
    
    
     
    
     
    
  </div>
</div>
`;

exports[`Unstake > Unstake operation > should render the unstake view 1`] = `
<div
  class="operation"
>
  <div
    class="dusk-wizard"
  >
    <div
      slot="stepper"
    >
      <div
        class="dusk-stepper dusk-stepper--variant--primary"
        style="--columns: 3; --progress-width: 0%;"
      >
        <span
          aria-current="step"
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step dusk-stepper__step--processed"
        >
          1
        </span>
         
        <span
          class="dusk-stepper__step-label"
          id="step-some-generated-id"
        >
          Amount
           
        </span>
        <span
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step"
        >
          2
        </span>
         
        <span
          class="dusk-stepper__step-label dusk-stepper__step-label--invisible"
          id="step-some-generated-id"
        >
          Overview
           
        </span>
        <span
          aria-labelledby="step-some-generated-id"
          class="dusk-stepper__step"
        >
          3
        </span>
         
        <span
          class="dusk-stepper__step-label dusk-stepper__step-label--invisible"
          id="step-some-generated-id"
        >
          Done
           
        </span>
        
      </div>
      
    </div>
     
     
     
    <div
      class="operation__amount-wrapper svelte-1p653d4"
    >
      <p>
        Amount:
      </p>
       
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          USE MAX
        </span>
      </button>
    </div>
     
    <div
      class="operation__input-wrapper svelte-1p653d4"
    >
      <input
        class="dusk-textbox dusk-textbox-number operation__input-field"
        max="278"
        min="false"
        required=""
        step="0.000000001"
        type="number"
      />
      
       
      <svg
        class="dusk-icon dusk-icon--size--default"
        data-tooltip-id="main-tooltip"
        data-tooltip-text="DUSK"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
        />
      </svg>
      
    </div>
     
    <div
      class="gas-settings svelte-1daay07"
    >
      <dl
        class="gas-settings__edit svelte-1daay07"
      >
        <div
          class="gas-fee svelte-1c6b481"
        >
          <span>
            Fee:
          </span>
           
          <div
            class="gas-fee__amount svelte-1c6b481"
          >
            <div
              class="gas-fee__amount-value svelte-1c6b481"
            >
              <span
                class="svelte-1c6b481"
              >
                max
              </span>
               
              <span
                class="svelte-1c6b481"
              >
                0.020000000
              </span>
            </div>
             
            <svg
              class="dusk-icon dusk-icon--size--default dusk-amount__icon"
              data-tooltip-id="main-tooltip"
              data-tooltip-text="DUSK"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
              />
            </svg>
            
          </div>
        </div>
        <dd>
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
            type="button"
          >
            <span
              class="dusk-button__text"
            >
              EDIT
            </span>
          </button>
        </dd>
      </dl>
       
    </div>
     
    
     
    <div
      class="dusk-wizard__step-navigation"
    >
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
        type="button"
      >
        <svg
          class="dusk-icon dusk-icon--size--default dusk-button__icon"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
          />
        </svg>
        
         
        <span
          class="dusk-button__text"
        >
          Back
        </span>
        
      </button>
      
       
      <button
        class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
        type="button"
      >
        <span
          class="dusk-button__text"
        >
          Next
        </span>
         
        <svg
          class="dusk-icon dusk-icon--size--default dusk-button__icon"
          role="graphics-symbol"
          viewBox="0 0 24 24"
        >
          <path
            d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
          />
        </svg>
        
      </button>
    </div>
    
    
     
    
     
    
  </div>
</div>
`;
